File: csharp/Exo1/Matcher.cs
using System;

namespace Soat.CleanCoders.Refactoring.Exo1
{
    public class Matcher
    {
        public bool Match(int[] expected, int[] actual, int clipLimit, int delta)
        {
            //Clip "too large" values
            for (var i = 0; i < actual.Length; i++)
            {
                if (actual[i] > clipLimit)
                {
                    actual[i] = clipLimit;
                }
            }

            //Check for length differences
            if (actual.Length != expected.Length)
            {
                return false;
            }

            //Check that each entry within the expected +/- delta
            for (var i = 0; i < expected.Length; i++)
            {
                if (Math.Abs(expected[i] - actual[i]) > delta)
                {
                    return false;
                }
            }

            return true;
        }
    }
}

File: javascript/app/exo1/matcher.js
"use strict"

class Matcher {

    constructor(clipLimit, delta){
        this.clipLimit = clipLimit;
        this.delta = delta;
    }


    match(expected, actual){
        actual = this.setClipLimit(actual, this.clipLimit);

        //Check for length differences
        if (!this.isEqual(actual.length, expected.length))
            return false;

        //Check that each entry within the expected +/- delta
        for (let i = 0; i < expected.length; i++) {
            let nbAbs = Math.abs(expected[i] - actual[i]);
            if (this.isSuperior(nbAbs, this.delta)){
                return false;
            }

        }
        return true;
    }

    isSuperior(nbAbs, delta) {
        return nbAbs > delta;
    }

    isEqual(actual, expected) {
        return actual == expected;
    }

    setClipLimit(tab, clipLimit) {
        for (let i = 0; i < tab.length; i++){
            if (this.isSuperior(tab[i], clipLimit)){
                tab[i] = clipLimit;
            }
        }
        return tab;
    }
}
module.exports = Matcher;